src/charm.py

```diff
@@ -1,427 +1,269 @@
+        """Unpacks and returns the object-storage relation data.
 
-        URL generated only if 'service' is set, otherwise it returns empty string.
+        Raises CheckFailedError if an anticipated error occurs.
         """
-        endpoint_url = ""
-        if "service" in obj_storage and len(obj_storage["service"]) > 0:
-            endpoint_url = f"http://{obj_storage['service']}"
-            if "namespace" in obj_storage and len(obj_storage["namespace"]) > 0:
-                endpoint_url += f".{obj_storage['namespace']}"
-            if "port" in obj_storage and len(obj_storage["port"]) > 0:
-                endpoint_url += f":{obj_storage['port']}"
-
-        return endpoint_url
-
-
-class CheckFailedError(Exception):
-    """Raise this exception if one of the checks in main fails."""
-
-    def __init__(self, msg, status_type=StatusBase):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
-
-
-def _gen_obj_storage_endpoint_url(obj_storage):
-    """Generate object storage endpoint URL."""
-    return f"http://{obj_storage['service']}.{obj_storage['namespace']}:{obj_storage['port']}"
-
-
-def _b64_encode_dict(d):
-    """Returns the dict with values being base64 encoded."""
-    # Why do we encode and decode in utf-8 first?
-    return {k: b64encode(v.encode("utf-8")).decode("utf-8") for k, v in d.items()}
-
-
-def _minio_credentials_dict(obj_storage):
-    """Returns a dict of minio credentials with the values base64 encoded."""
-    minio_credentials = {
-        "AWS_ENDPOINT_URL": _gen_obj_storage_endpoint_url(obj_storage),
-        "AWS_ACCESS_KEY_ID": obj_storage["access-key"],
-        "AWS_SECRET_ACCESS_KEY": obj_storage["secret-key"],
-        "USE_SSL": str(obj_storage["secure"]).lower(),
-    }
-    return _b64_encode_dict(minio_credentials)
-
-
-def _seldon_credentials_dict(obj_storage):
-    """Returns a dict of seldon init-container object storage credentials, base64 encoded."""
-    credentials = {
-        "RCLONE_CONFIG_S3_TYPE": "s3",
-        "RCLONE_CONFIG_S3_PROVIDER": "minio",
-        "RCLONE_CONFIG_S3_ACCESS_KEY_ID": obj_storage["access-key"],
-        "RCLONE_CONFIG_S3_SECRET_ACCESS_KEY": obj_storage["secret-key"],
-        "RCLONE_CONFIG_S3_ENDPOINT": _gen_obj_storage_endpoint_url(obj_storage),
-        "RCLONE_CONFIG_S3_ENV_AUTH": "false",
-    }
-    return _b64_encode_dict(credentials)
+        if not ((obj_storage := interfaces["object-storage"]) and obj_storage.get_data()):
+            raise ErrorWithStatus("Waiting for object-storage relation data", WaitingStatus)
 
+        try:
+            obj_storage = list(obj_storage.get_data().values())[0]
+        except Exception as e:
+            raise ErrorWithStatus(
+                f"Unexpected error unpacking object storage data - data format not "
+                f"as expected. Caught exception: '{str(e)}'",
+                BlockedStatus,
+            )
```

Comment:
This sounds like an Error rather than Blocked.  Unless it can happen in regular operation of the charm, and the charm can later resolve itself

Reply:
Blocked is consistent with rest of our rewrites .... do we want to change it in every sidecar charm? If so Lets keep this Blocked and create a task. What do you think?

Reply:
Agreed